src/annotator.cpp

```diff
@@ -0,0 +1,361 @@
+        }
+        for (size_t i = 0; i < model->units(u)->unitCount(); ++i) {
+            std::string prefix;
+            std::string reference;
+            double exponent;
+            double multiplier;
+            model->units(u)->unitAttributes(i, reference, prefix, exponent, multiplier, id);
+            if (!id.empty()) {
+                idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::UNIT, std::make_pair(model->units(u), i))));
+            }
+        }
+        if (model->units(u)->isImport() && model->units(u)->importSource() != nullptr) {
+            id = model->units(u)->importSource()->id();
+            idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::IMPORT, model->units(u)->importSource())));
+        }
+    }
+    // Components.
+    for (size_t c = 0; c < model->componentCount(); ++c) {
+        listComponentIdsAndItems(model->component(c), idList);
+    }
+    // Encapsulation.
+    id = model->encapsulationId();
+    if (!id.empty()) {
+        idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::ENCAPSULATION, std::string())));
+    }
+
+    return idList;
+}
+
+void Annotator::build(const ModelPtr &model)
+{
+    mPimpl->isBuilt = false;
+    removeAllIssues();
+    mPimpl->mIdList.clear();
+    mPimpl->mIdList = listIdsAndItems(model);
+    mPimpl->isBuilt = true;
+    mPimpl->mModel = model;
+}
+
+AnyItem Annotator::item(const std::string &id)
+{
+    AnyItem item;
+    if (!mPimpl->isBuilt) {
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("Please call the Annotator::build function before attempting to access items by their id.");
+        issue->setLevel(libcellml::Issue::Level::ERROR);
+        addIssue(issue);
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        return item;
+    }
+    if (mPimpl->mIdList.empty()) {
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("The id map is empty; the supplied model has no id attributes.");
+        issue->setLevel(libcellml::Issue::Level::ERROR);
+        addIssue(issue);
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        return item;
+    }
+
+    // Retrieve the item from the idList.
+    ItemList::iterator it;
+    it = mPimpl->mIdList.find(id);
+
+    if (it == mPimpl->mIdList.end()) {
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("Could not find an item with id='" + id + "' in the model.");
+        issue->setLevel(libcellml::Issue::Level::WARNING);
+        addIssue(issue);
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        // Adding the requested id string and this issue to the map, so that future requests (ie: to the .second)
+        // will return the same IssuePtr rather than creating a new one.
+        mPimpl->mIdList.insert(std::make_pair(id, item));
+        return item;
+    }
+    return mPimpl->mIdList[id];
```

Comment:
In most (all?) methods we have where the user requests an object by "name", you get a `nullptr` if that name doesn't exist or an object if it does. Can we not do that here?

Reply:
Alas, no, for two reasons.  I tried this originally, and I couldn't cast a nullptr into the required return type that the compiler would accept.  Given that, it seemed better to return something that would be helpful to the user, and prevent duplicate issues being generated (as above).

Reply:
If they request a specific item type they will get a `nullptr` (eg: annotator->component("iDontExist") will return `nullptr`.  It's only the `item` function that is different.